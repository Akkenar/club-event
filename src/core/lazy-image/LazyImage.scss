.LazyImage {
  margin: 2em 0;
}

.LazyImage__Image {
  width: 100%;
  height: auto;
}

.LazyImage__Legend {
  font-style: italic;
}
