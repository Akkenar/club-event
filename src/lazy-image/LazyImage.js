import React from 'react';
import Image from './Image';
import Placeholder from './Placeholder';

const OBSERVER_OPTIONS = {
  rootMargin: '0px 0px 0px 50px',
};

class LazyImage extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      isDisplayed: false,
    };

    this.observer = new IntersectionObserver(
      this.handleIntersect.bind(this),
      OBSERVER_OPTIONS
    );

    this.handleRef = this.handleRef.bind(this);
  }

  handleIntersect(entries) {
    entries.forEach(entry => {
      if (entry.isIntersecting) {
        this.setState({
          isDisplayed: true,
        });

        if (this.observer) {
          this.observer.unobserve(entry.target);
        }
      }
    });
  }

  handleRef(element) {
    if (!element) {
      return;
    }

    // Don't bother observing anything if the image is already displayed.
    const { isDisplayed } = this.state;
    if (isDisplayed) {
      return;
    }

    this.observer.observe(element);
  }

  componentWillUnmount() {
    if (this.observer) {
      this.observer.disconnect();
    }
  }

  render() {
    const { isDisplayed } = this.state;
    const { src, srcwebp, width, height, alt } = this.props;

    if (!isDisplayed) {
      return (
        <Placeholder
          src={src}
          height={height}
          width={width}
          alt={alt}
          handleRef={this.handleRef}
        />
      );
    }

    return (
      <Image
        src={src}
        srcwebp={srcwebp}
        height={height}
        width={width}
        alt={alt}
      />
    );
  }
}

export default LazyImage;
