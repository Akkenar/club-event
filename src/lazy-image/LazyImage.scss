.LazyImage {
  width: 100%;
  height: auto;
}
