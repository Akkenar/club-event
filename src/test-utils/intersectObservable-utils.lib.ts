import { mockGlobalProperty } from './test-utils.lib';

const allObservers: IntersectionObserverMock[] = [];

class IntersectionObserverMock {
  private readonly cb: any;
  public constructor(cb: any) {
    this.cb = cb;

    // To allow global triggering.
    allObservers.push(this);
  }

  public trigger() {
    this.cb([{ isIntersecting: true }]);
  }

  public observe() {
    // Nothing
  }

  public unobserve() {
    // Nothing.
  }

  public disconnect() {
    // Nothing.
  }
}

class IntersectionObserverEntryMock {
  public intersectionRatio() {
    // Nothing.
  }

  public isIntersecting() {
    // Nothing.
  }
}

export function mockInterestObservable(): () => void {
  // Reset all the mocks
  allObservers.slice(0);

  mockGlobalProperty(window)('IntersectionObserver')(IntersectionObserverMock);
  mockGlobalProperty(window)('IntersectionObserverEntry')(
    IntersectionObserverEntryMock,
  );

  // Trigger all.
  return () => {
    allObservers.forEach((observer: IntersectionObserverMock) => {
      observer.trigger();
    });
  };
}
