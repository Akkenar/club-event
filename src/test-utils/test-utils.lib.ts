import { fireEvent } from '@testing-library/react';

export function mockGrecaptcha() {
  (window as any).grecaptcha = {
    getResponse: () => 'response',
    render: jest.fn(),
    reset: jest.fn(),
  };
}

export function mockResponse(response: any) {
  const mock = typeof response === 'string' ? { result: response } : response;

  // Mock the fetch data as JSDom doesn't implement fetch.
  // @ts-ignore
  window.fetch = jest.fn(() =>
    Promise.resolve({
      text: () =>
        new Promise(resolve => {
          const mockResult = JSON.stringify(mock);
          setTimeout(() => {
            resolve(mockResult);
          }, 50);
        }),
    }),
  );
}

export function setValidInputData(
  validData: any,
  { getByTestId }: { getByTestId: any },
) {
  Object.keys(validData).forEach((key: string) => {
    fireEvent.change(getByTestId(key), {
      // @ts-ignore
      target: { value: validData[key] },
    });
  });
}

export const mockGlobalProperty = (globalObject: any) => (key: string) => (
  value: any,
) => {
  const original = globalObject[key];
  Object.defineProperty(globalObject, key, { value, writable: true });
  return () => mockGlobalProperty(globalObject)(key)(original);
};

export function forceLoadImage() {
  Object.defineProperty(Image.prototype, 'onload', {
    configurable: true,
    set(fn: () => void) {
      // Call the function directly to simulate the loading
      fn();
    },
  });
}
