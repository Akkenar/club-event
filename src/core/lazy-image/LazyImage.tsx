import * as React from 'react';
import Image from './Image';

import './LazyImage.scss';
import Placeholder from './Placeholder';
import { useIntersectionObserver } from './useIntersectionObserver';

interface LazyImageProps {
  src: any;
  srcwebp: any;
  width?: number;
  height?: number;
  alt: string;
  className?: string;
}

const LazyImage = ({
  src,
  srcwebp,
  width,
  height,
  alt,
  className,
}: LazyImageProps) => {
  const [isDisplayed, handleRef] = useIntersectionObserver();

  if (!handleRef) {
    // While the observer is created.
    return null;
  }

  if (!isDisplayed) {
    return (
      <Placeholder
        className={`LazyImage ${className || ''}`}
        src={src}
        height={height}
        width={width}
        alt={alt}
        handleRef={handleRef}
      />
    );
  }

  return (
    <Image
      className={`LazyImage no-print ${className || ''}`}
      src={src}
      srcwebp={srcwebp}
      height={height}
      width={width}
      alt={alt}
    />
  );
};

export default LazyImage;
