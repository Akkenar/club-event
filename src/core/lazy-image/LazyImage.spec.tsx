import * as React from 'react';
import { render, waitForElement } from 'react-testing-library';
import { mockGlobalProperty } from '../../test-utils/test-utils.lib';
import LazyImage from './LazyImage';

const fakeElements = [{ isIntersecting: true }, { isIntersecting: false }];

/* tslint:disable */
class IntersectionObserverMock {
  constructor(cb: any) {
    // Hack to invoke the triggering of the intersection when we want it.
    setTimeout(() => cb(fakeElements), 100);
  }

  observe() {
    // Nothing.
  }

  unobserve() {
    // Nothing.
  }

  disconnect() {
    // Nothing.
  }
}

class IntersectionObserverEntryMock {
  intersectionRatio() {
    // Nothing.
  }

  isIntersecting() {
    // Nothing.
  }
}
/* tslint:enable */

describe('LazyImage', () => {
  beforeEach(() => {
    mockGlobalProperty(window)('IntersectionObserver')(
      IntersectionObserverMock
    );
    mockGlobalProperty(window)('IntersectionObserverEntry')(
      IntersectionObserverEntryMock
    );
  });

  it('should match snapshot with placeholder', async () => {
    const wrapper = render(
      <LazyImage src="foo" srcwebp="bar" alt="al" className="test" />
    );
    await waitForElement(() => wrapper.getByTestId('lazyloaded-placeholder'));
    expect(wrapper.baseElement).toMatchSnapshot();
  });

  it('should match snapshot with image', async () => {
    const wrapper = render(
      <LazyImage src="foo" srcwebp="bar" alt="alt" className="test" />
    );
    await waitForElement(() => wrapper.getByTestId('lazyloaded-image'));
    expect(wrapper.baseElement).toMatchSnapshot();
  });
});
