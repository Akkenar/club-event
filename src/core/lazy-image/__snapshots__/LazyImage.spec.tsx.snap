// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LazyImage should match snapshot with image 1`] = `
<body>
  <div>
    <picture>
      <source
        srcset="bar"
        type="image/webp"
      />
      <img
        alt="alt"
        class="LazyImage no-print test"
        data-testid="lazyloaded-image"
        src="foo"
        title="alt"
      />
    </picture>
  </div>
</body>
`;

exports[`LazyImage should match snapshot with placeholder 1`] = `
<body>
  <div>
    <svg
      aria-label="al"
      class="LazyImage test"
      data-testid="lazyloaded-placeholder"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d=""
      />
    </svg>
  </div>
</body>
`;
